source/extensions/transport_sockets/internal_upstream/config.cc

```diff
@@ -0,0 +1,180 @@
+    return std::make_unique<InternalSocketFactory>(context, outer_config,
+                                                   std::move(inner_transport_factory));
+  }
+};
+
+} // namespace
+
+Config::Config(
+    const envoy::extensions::transport_sockets::internal_upstream::v3::InternalUpstreamTransport&
+        config_proto,
+    Stats::Scope& scope)
+    : stats_{ALL_INTERNAL_UPSTREAM_STATS(POOL_COUNTER_PREFIX(scope, "internal_upstream."))} {
+  for (const auto& metadata : config_proto.passthrough_metadata()) {
+    MetadataKind kind;
+    switch (metadata.kind().kind_case()) {
+    case envoy::type::metadata::v3::MetadataKind::KindCase::kHost:
+      kind = MetadataKind::Host;
+      break;
+    case envoy::type::metadata::v3::MetadataKind::KindCase::kCluster:
+      kind = MetadataKind::Cluster;
+      break;
+    default:
+      throw EnvoyException(
+          absl::StrCat("metadata type is not supported: ", metadata.kind().DebugString()));
+    }
+    metadata_sources_.push_back(MetadataSource(kind, metadata.name()));
+  }
+  for (const auto& filter_state_object : config_proto.passthrough_filter_state_objects()) {
+    filter_state_names_.push_back(filter_state_object.name());
+  }
+}
+std::unique_ptr<envoy::config::core::v3::Metadata>
+Config::extractMetadata(const Upstream::HostDescriptionConstSharedPtr& host) const {
+  if (metadata_sources_.empty()) {
+    return nullptr;
+  }
+  auto metadata = std::make_unique<envoy::config::core::v3::Metadata>();
+  for (const auto& source : metadata_sources_) {
+    OptRef<const envoy::config::core::v3::Metadata> source_metadata;
+    switch (source.kind_) {
+    case MetadataKind::Host: {
+      if (host->metadata()) {
+        source_metadata = makeOptRef(*host->metadata());
+      }
+      break;
+    }
+    case MetadataKind::Cluster:
+      source_metadata = makeOptRef(host->cluster().metadata());
+      break;
+    }
+    if (source_metadata && source_metadata->filter_metadata().contains(source.name_)) {
+      (*metadata->mutable_filter_metadata())[source.name_] =
+          source_metadata->filter_metadata().at(source.name_);
+    } else {
+      ENVOY_LOG(trace, "Internal upstream missing metadata: {}", source.name_);
+      stats_.no_metadata_.inc();
+    }
+  }
+  return metadata;
+}
+
+std::unique_ptr<IoSocket::UserSpace::FilterStateObjects>
+Config::extractFilterState(const StreamInfo::FilterStateSharedPtr& filter_state) const {
+  if (filter_state_names_.empty()) {
+    return nullptr;
+  }
+  auto filter_state_objects = std::make_unique<IoSocket::UserSpace::FilterStateObjects>();
+  for (const auto& name : filter_state_names_) {
+    try {
+      auto object = filter_state->getDataSharedMutableGeneric(name);
+      if (object == nullptr) {
+        ENVOY_LOG(trace, "Internal upstream missing filter state: {}", name);
+        stats_.no_filter_state_.inc();
+        continue;
+      }
+      filter_state_objects->emplace_back(name, object);
```

Comment:
Is it possible to not copy those filter state data and just pass the filter_state reference to the downstream listener? since we invoke the downstream listener's onAccept immediately when create the client connection (even we didn't return the client connection yet), 
https://github.com/envoyproxy/envoy/blob/006bbc3614724ead239fcc3a2438b4dd6b9173e6/source/extensions/bootstrap/internal_listener/client_connection_factory.cc#L66

I'm thinking is that means the upstream stream info and state_filter are still alive and have no chance to be released.

Reply:
What might be confusing - this is downstream stream info, not upstream stream info. I didn't think about transfer from downstream to upstream, which also makes sense, but is not needed here.

Reply:
What I'm thinking is utilizing the special part of the internal connection. When the downstream creates the internal `Network::ClientConnectionImpl` by `InternalClientConnectionFactory::createClientConnection`, it will invoke the upstream's `onAccept` callback directly inside `InternalClientConnectionFactory::createClientConnection`, 
https://github.com/envoyproxy/envoy/blob/006bbc3614724ead239fcc3a2438b4dd6b9173e6/source/extensions/bootstrap/internal_listener/client_connection_factory.cc#L66

so I'm thinking that means the downstream and upstream lifetime already bind together. When you do the merge with upstream filter states in the InternalListener's `onAccept` callback, the downstream filter stats shouldn't have chance to end the life, because the `InternalClientConnectionFactory::createClientConnection` method call event not finish yet. (onAccept call is the last line of this method)

https://github.com/envoyproxy/envoy/pull/19435/files#diff-e0f716749375fc91f21d624aca995ad6a0af4147666fee56b9703e8774e2dc83R64

But I didn't validate it, I could be wrong.